src/ILVerification/src/TypeVerifier.cs

```diff
@@ -0,0 +1,97 @@
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeVerifier
+{
+    internal class TypeVerifier
+    {
+        private EcmaModule _module;
+        private readonly TypeDefinitionHandle _typeDefinitionHandle;
+        private readonly Lazy<ResourceManager> _stringResourceManager;
+        private static readonly string _errorPrefix = "[MD]: Error: ";
+
+        public Action<ErrorArgument[], VerifierError, string, object[]> ReportVerificationError
+        {
+            set;
+            private get;
+        }
+
+        public TypeVerifier(EcmaModule module, TypeDefinitionHandle typeDefinitionHandle, Lazy<ResourceManager> stringResourceManager)
+        {
+            _module = module;
+            _typeDefinitionHandle = typeDefinitionHandle;
+            _stringResourceManager = stringResourceManager;
+        }
+
+        public void Verify()
+        {
+            VerifyInterfaces();
+        }
+
+        public void VerifyInterfaces()
+        {
+            TypeDefinition typeDefinition = _module.MetadataReader.GetTypeDefinition(_typeDefinitionHandle);
+            EcmaType type = (EcmaType)_module.GetType(_typeDefinitionHandle);
+
+            if (type.IsInterface)
+            {
+                return;
+            }
+
+            InterfaceImplementationHandleCollection interfaceHandles = typeDefinition.GetInterfaceImplementations();
+            int count = interfaceHandles.Count;
+            if (count == 0)
+            {
+                return;
+            }
+
+            // Look for duplicates.
+            List<InterfaceMetadataObjects> implementedInterfaces = new List<InterfaceMetadataObjects>();
+            foreach (InterfaceImplementationHandle interfaceHandle in interfaceHandles)
+            {
+                InterfaceImplementation interfaceImplementation = _module.MetadataReader.GetInterfaceImplementation(interfaceHandle);
+                DefType interfaceType = _module.GetType(interfaceImplementation.Interface) as DefType;
+                if (interfaceType == null)
+                {
+                    ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadBadFormat, type);
+                }
+
+                InterfaceMetadataObjects imo = new InterfaceMetadataObjects
+                {
+                    DefType = interfaceType,
+                    InterfaceImplementationHandle = interfaceHandle
+                };
+
+                if (!implementedInterfaces.Contains(imo))
+                {
+                    implementedInterfaces.Add(imo);
+                }
+                else
+                {
+                    string message = _stringResourceManager.Value.GetString(VerifierError.InterfaceImplHasDuplicate.ToString(), CultureInfo.InvariantCulture);
+                    ReportVerificationError(null, VerifierError.InterfaceImplHasDuplicate, $"{_errorPrefix}{message}", new object[] { type.ToString(), interfaceType.ToString(), _module.MetadataReader.GetToken(interfaceHandle) });
```

Comment:
`ReportVerificationError` is a delegate...but maybe is too specific to pass `interfaceType`, if here we put all Type check, isn't it? We won't have an interfaceType object to pass for all check, we should have ever errorCode, formatString, args, an helper like:
```cs
private void RaiseReportVerificationError(VerifierError error, object[] args)
{
   string message = _stringResourceManager.Value.GetString(error.ToString(), CultureInfo.InvariantCulture);
   ReportVerificationError(error, $"{_errorPrefix}{message}", args);
}
...
else
{
   RaiseReportVerificationError(VerifierError.InterfaceImplHasDuplicate, new object[] { type, interfaceType, _module.MetadataReader.GetToken(interfaceHandle) });
}
...
```
And on VerifyType callback
```cs
typeVerifier.ReportVerificationError = (code, message, args) =>
  {
    builder.Add(new VerificationResult()
    {
       Code = code,
       Message = message,
       Args = args
    });
};
```
What do you think?Have I misunderstood?

Reply:
Yes, something like this. 

- Make `RaiseReportVerificationError` shorter since it will be repeated a lot. Can it be just `VerificationError` ?
- Do you really need to pass `_stringResourceManager` to `TypeVerifier`? Can it just stay inside `Verifier.cs` ?

Reply:
> Make RaiseReportVerificationError shorter since it will be repeated a lot. Can it be just VerificationError ?

OK

> Do you really need to pass _stringResourceManager to TypeVerifier? Can it just stay inside Verifier.cs ?

No I can change signature of delegate 

`public Action<VerifierError,object[]> ReportVerificationError`

and concat prefix  "[MD]: Error: " on Verifier.cs because I don't like to have repeated prefix in resources, what about it?

Reply:
sounds good

Reply:
code updated...last thing that I don't understand(and I don't know if is useful) is how to get this token https://github.com/lewischeng-ms/sscli/blob/master/clr/src/tools/peverify/main.cpp#L324 as I said on PR description. Here seems set to 0 https://github.com/lewischeng-ms/sscli/blob/master/clr/src/md/compiler/mdvalidator.cpp#L515

Reply:
Isn't it the token that corresponds to `typeDefinitionHandle` ?

Reply:
That's what I thought...but if I run PEVerify on similar tests(I created some PEVerifiable dll to compare output) token values are very close `[MD]: Error: InterfaceImpl has a duplicate, token=0x09000002. [token:0x09000001]`
```il
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

.assembly InterfaceDefinition
{
}

.assembly extern System.Runtime
{
}

.class interface public auto ansi abstract Interface
{	
	.method public hidebysig newslot abstract virtual 
		instance void M1 () cil managed 
	{
	}

	.method public hidebysig newslot abstract virtual 
		instance void M2 () cil managed 
	{
	}

}

.class public auto ansi beforefieldinit Class
	extends [System.Runtime]System.Object
	implements Interface, Interface
{	
	.method public final hidebysig newslot virtual 
		instance void M1 () cil managed 
	{		
		.maxstack 8

		IL_0000: ret
	}

	.method public hidebysig specialname rtspecialname 
		instance void .ctor () cil managed 
	{		
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: call instance void [System.Runtime]System.Object::.ctor()
		IL_0006: ret
	}
} 

```
The only difference is that this il contains interface and class...maybe this is the reason for close values. Another strange thing is that interface token value is smaller than ipotetic `type` token, maybe related to physical layout. I preferred to ask here because maybe this value is important and it's not so clear to me where on PEVerify codebase this value is set. I'm currently studying ECMA spec and this question could be silly, sorry.